### ISSUE TITLE
WiPy: REPL lock-up after file read & list operations

### ISSUE THREAD

**Issue by robert-hh** [micropython/micropython#2548]
This little piece of code runs into a OSError or Memory Error:

```
import gc
def get_file(fname):
    content = []
    with open(fname) as f:
        for l in f:
            content.append(l)
    return content

t = get_file("getfile.py")
print("This is the first verse")
gc.collect()
print("And this is the second verse")
```

The first print statement is executed, but then OSError comes up, and after that, the REPL prompt is affected too.
Wipy, today's build:
MicroPython v1.8.5-27-g979e9a4 on 2016-10-22; WiPy with CC3200
P.S.: I know that I could use f.readlines(), but then it crashes at another place of the larger code, this was taken from. I have a lost of list handling in that code, and that runs well, as long as no file operations are involved.
This issue is not present in the 1.8.5 version of Oct. 17
Update: the bug shows up at the transition from build 17ba6ef -> 84679e

**Comment by dpgeorge** [comment]
Thanks for the bug report.  The problem was that `file.close()` was being called multiple times and only the first time succeeded, the rest failed with OSError.  Furthermore, the finaliser `file.__del__()` was being called from the `gc.collect()` and raising an exception within the garbage collection run.  This lead to the heap remaining locked at the REPL, hence the MEmoryError.  This issue also affected stmhal (and probably esp8266).

So, in 56942019309645781d330312f5944db2d4cb5cd7 I fixed file.close() so it doesn't try to close the file more than once.  It fixes the problem described above.

BUT, there seems to be a separate problem with cc3200 (wipy).  If the code example in the OP above is saved to a file "getfile.py", then the following is run at the REPL, it eventually crashes (REPL becomes unresponsive):
```
>>> import getfile
>>> getfile.get_file('getfile.py')
>>> getfile.gc.collect()
[use up/down arrow keys to go through history]
[use tab completion]
>>> getfile.get_file('getfile.py')
>>> getfile.gc.collect()
[continue doing these things]
```
I checked that this bug is there as far back as v1.8.  I don't have resources to fix this so will defer to @danicampora .
